SimTKcommon/src/Pathname.cpp

```diff
@@ -218,13 +220,187 @@ void Pathname::deconstructPathname( const string&   name,
     }
 
     // Now we can put together the canonicalized directory.
-    if (isAbsolutePath) {
+    if (dontApplySearchPath) {
         if (!drive.empty())
             directory = drive + ":";
         directory += "/";
     }
 
-    for (int i = (int)segmentsInReverse.size()-1; i >= 0; --i)
+    for (int i = (int)segmentsInReverse.size() - 1; i >= 0; --i)
+        directory += segmentsInReverse[i] + "/";
+
+    // Fix the slashes.
+    makeNativeSlashesInPlace(directory);
+
+    // If there is a .extension, strip it off.
+    string::size_type lastDot = fileName.rfind('.');
+    if (lastDot != string::npos) {
+        extension = fileName.substr(lastDot);
+        fileName.erase(lastDot);
+    }
+}
+
+void Pathname::deconstructPathnameUsingSpecifiedWorkingDirectory(const std::string& swd,
+                                                                 const std::string& path,
+                                                                 std::string& directory,
+                                                                 std::string& fileName,
+                                                                 std::string& extension)
+{
+    directory.erase(); fileName.erase(); extension.erase();
+    string pathdrive, swddrive, finaldrive;
+    String processed;
+
+    // Remove all the white space and make all the slashes be forward ones.
+    // (For Windows they'll be changed to backslashes later.)
+    String pathcleaned = String::trimWhiteSpace(path).replaceAllChar('\\', '/');
+    if (pathcleaned.empty())
+        return; // path consisted only of white space
+    removeDriveInPlace(pathcleaned, pathdrive);
+
+    String swdcleaned = String::trimWhiteSpace(swd).replaceAllChar('\\', '/');
+    
+    // If swd was empty, then just call deconstructPathname() on the path.
+    if (swdcleaned.empty()) {
+        bool dontApplySearchPath;
```

Comment:
Rather than repeating this code, how about adding a new method `deconstructAbsolutePathname()` that does it? Then replace `getAbsolutePathname()` with:
```c++
static std::string getAbsolutePathname(const std::string& pathName) {
    std::string directory, fileName, extension;
    desconstructAbsolutePathname(pathName, directory, fileName, extension);
    return directory + fileName + extension;
}